PhabricatorBMOAuthProvider.php

```diff
@@ -0,0 +1,400 @@
+
+    $request = $controller->getRequest();
+
+    // When Bugzilla sends us a "back channel" POST during registration...
+    if ($request->isHTTPPost()) {
+      return $this->processLoginRequestBackChannelPost($controller, $request);
+    }
+
+    // The last step in Bugzilla's verification process
+    return $this->processLoginRequestConfirmationGet($controller, $request);
+  }
+
+  private function processLoginRequestBackChannelPost($controller) {
+    $config = $this->getConfig();
+    $account = null;
+    $response = null;
+
+    // Get the `client_api_key` and `client_api_login` from Bugzilla's POST data
+    $post_info = array();
+    try {
+      $post_info = phutil_json_decode(PhabricatorStartup::getRawInput());
+    }
+    catch(Exception $ex) {
+      $this->throwException(
+        pht('Phabricator BMO Authentication failed due to '.
+            'invalid JSON from Bugzilla.')
+      );
+    }
+
+    // Throw exception if either key is not provided by Bugzilla
+    if(!isset($post_info['client_api_key']) || !isset($post_info['client_api_login'])) {
+      $this->throwException(
+        pht('Phabricator BMO Authentication failed due to '.
+            'incomplete JSON from Bugzilla.')
+      );
+    }
+
+    // Generate a transaction code which we'll receive back from Bugzilla
+    // To confirm the API and Client Login information which we saved
+    // In PHP7 this becomes: bin2hex(random_bytes(self::TRANSACTION_CODE_LENGTH))
+    $trans_code = bin2hex(openssl_random_pseudo_bytes(
+      $config->getProperty(self::CONFIG_KEY_TRANSACTION_CODE_LENGTH)
+    ));
+
+    $csrf = $controller->getExtraURIData();
+
+    // Create a temporary auth token to save the user info JSON provided
+    // by the POST from Bugzilla.  Implicitly validates CSRF.
+    $unguarded = AphrontWriteGuard::beginScopedUnguardedWrites();
+    id(new PhabricatorAuthTemporaryToken())
+      ->setTokenResource($trans_code)
+      ->setTokenCode($csrf)
+      ->setTokenType(self::TOKEN_TYPE)
+      ->setTokenExpires(time() + phutil_units('1 hour in seconds'))
+      ->setTemporaryTokenProperty('api_key', $post_info['client_api_key'])
+      ->setTemporaryTokenProperty('client_login', $post_info['client_api_login'])
+      ->save();
+    unset($unguarded);
+
+    // Return a JSON object with a `result` key
+    // This result represents the transaction key we'll use to look up
+    // the JSON information in the next phase of requests
+    $json_obj = array('result' => $trans_code);
+
+    // Render the response JSON
+    $response = id(new AphrontJSONResponse())
+      ->setAddJSONShield(false)
+      ->setHTTPResponseCode(200)
+      ->setContent($json_obj);
+    return array($account, $response);
+  }
+
+  private function processLoginRequestConfirmationGet($controller, $request) {
+    $response = null;
+
+    // Verify with CSRF as an additional security measure
+    $this->verifyAuthCSRFCode($request, $controller->getExtraURIData());
```

Comment:
I was initially scared by this because the `$controller->getExtraURIData()` is coming from the URL, which can be modified, but that value is compared to the `phcid` here:  https://github.com/phacility/phabricator/blob/3d816e94dfbedd5b9bdea0d9d83674a97bef0651/src/applications/auth/provider/PhabricatorAuthProvider.php#L465

Thus they need to actually have this session ID, thus this code is safe.